proposals/p2240.md

```diff
@@ -0,0 +1,758 @@
+fit for Carbon. Consider this example:
+
+```
+fn SumInts(..., [:]params: [i64;]) -> i64
+```
+
+If `[:]` and `...,` were suffix operators, this would instead be:
+
+```
+fn SumInts(params: [i64;][:] ...,) -> i64
+```
+
+But this is ambiguous about whether the `[:]` operator applies to
+`params: [i64;]` or just `[i64;]`, so we need to add a set of parentheses:
+
+```
+fn SumInts((params: [i64;])[:] ...,) -> i64
+```
+
+To state the point more generally: when an expression mixes prefix and suffix
+operators, their order is determined by precedence and parentheses, which is
+usually less readable than relying on the linear order of operators with the
+same fixity. The binding operator `:` is effectively a prefix operator (because
+its left-hand side is a single token), so if `[:]` is a suffix operator, code
+that mixes the two will be somewhat harder to read. That's a particular problem
+because, as our motivating examples illustrate, variadic function signatures
+will often have those two operators in close conjunction, and readability is at
+a premium in function signatures.
+
+C solved a very similar problem with mixing prefix `*` and postfix `.` by
+introducing a separate `->` operator. We could in principle solve this issue the
+same way, introducing a separate "variadic binding operator" `$`, such that
+`params$ [i64;]` is equivalent to `(params: [i64;])[:]`. However, this would
+trade a readability problem for a teachability problem, by introducing more
+syntax that programmers need to learn before they can even read variadic
+function signatures.
+
+### Keyword syntax
+
+> **TODO:** Decide whether to make this the primary proposal, and restructure
+> the discussion as appropriate.
+
+We could build the syntax around keywords rather than symbolic tokens, much like
+the
+[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
+For concreteness, the keywords I will discuss here are as follows:
+
+| Symbolic token | Keyword replacement               |
+| -------------- | --------------------------------- |
+| `[:]`          | `expand`                          |
+| `...,`         | `repeat` and `unpack` (see below) |
+| `...and`       | `all_of`                          |
+| `...or`        | `any_of`                          |
+| `...{`         | `repeat {`                        |
+
+This approach seems to work best in combination with some other nominally
+independent changes:
+
+-   Forbid `expand` in pattern contexts.
+-   Allow `repeat` only in expression contexts and `unpack` only in pattern
+    contexts.
+-   Require any usage of a pack-type variable to be prefixed with a new keyword
+    `each`.
+
+With these changes, the examples given earlier would read as follows:
+
+```carbon
+// Computes the sum of its arguments, which are i64s
+fn SumInts(unpack params: expand [i64;]) -> i64 {
+  var sum: i64 = 0;
+  repeat {
+    sum += each params;
+  }
+  return sum;
+}
```

Comment:
```suggestion
// Computes the sum of its arguments, which are i64s
fn SumInts(...each param: i64) -> i64 {
  var sum: i64 = 0;
  ...{
    sum += each param;
  }
  return sum;
}
```

Reply:
It's not clear to me that we need the braces in `...{...}`, in the case where we're expanding a single statement. I think we could handle:

```suggestion
// Computes the sum of its arguments, which are i64s
fn SumInts(...each param: i64) -> i64 {
  var sum: i64 = 0;
  ... sum += each param;
  return sum;
}
```

and potentially even

```carbon
fn Zip[...each ElementType:! type]
      (...each vector: Vector(each ElementType))
      -> Vector((...each ElementType)) {
  // Local pack.
  ...var each iter: auto = each vector.Begin();
  var result: Vector(...each ElementType);
  while (...and each iter != each vector.End()) {
    result.push_back((...*each iter));
    ...each iter++;
  }
  return result;
}
```

(There's an ambiguity here for a statement starting with `...{`, but it's easy to resolve with a single token of lookahead: `...{}` is invalid either way, `...{.` must be a single-statement expansion, and `...{` followed by anything else must be a block expansion.)

Reply:
Regarding omitting the braces, do you think it's important to resolve that now, or can we table it until the other syntax issues are settled?

Reply:
I think we can table it. It's interesting that this syntactic approach has some potentially nice future generalizations, but I don't think they're motivating enough to be a primary factor when picking the syntax.